examples/waku-chat/src/chat.rs

```diff
@@ -0,0 +1,90 @@
+use crate::network::*;
+use async_trait::async_trait;
+use overwatch::services::handle::ServiceStateHandle;
+use overwatch::services::relay::{NoMessage, Relay};
+use overwatch::services::state::{NoOperator, NoState};
+use overwatch::services::{ServiceCore, ServiceData, ServiceId};
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc::channel;
+
+/// Chat service handler
+/// displays received messages, send new ones
+pub struct ChatService {
+    service_state: ServiceStateHandle<Self>,
+}
+
+#[derive(Deserialize, Serialize)]
+struct Message {
+    user: usize,
+    msg: Box<[u8]>,
+}
+
+impl ServiceData for ChatService {
+    const SERVICE_ID: ServiceId = "Chat";
+    type Settings = usize;
+    type State = NoState<Self::Settings>;
+    type StateOperator = NoOperator<Self::State>;
+    type Message = NoMessage;
+}
+
+#[async_trait]
+impl ServiceCore for ChatService {
+    fn init(service_state: ServiceStateHandle<Self>) -> Self {
+        Self { service_state }
+    }
+
+    async fn run(self) {
+        let Self {
+            mut service_state, ..
+        } = self;
+        // TODO: waku should not end up in the public interface of the network service, at least not as a type
+        let mut network_relay: Relay<NetworkService<waku::Waku>> =
+            service_state.overwatch_handle.relay();
+        let user = service_state.settings_reader.get_updated_settings();
+        network_relay.connect().await.unwrap();
+        let (sender, mut receiver) = channel(1);
+        // TODO: typestate so I can't call send if it's not connected
+        network_relay
+            .send(NetworkMsg::Subscribe {
+                kind: EventKind::Message,
+                sender,
+            })
+            .await
+            .unwrap();
+
+        // send new messages
+        // for interactive stdin I/O it's recommended to
+        // use an external thread, see https://docs.rs/tokio/latest/tokio/io/struct.Stdin.html
+        std::thread::spawn(move || loop {
+            let mut input = String::new();
+            std::io::stdin()
+                .read_line(&mut input)
+                .expect("error reading message");
+            input.truncate(input.trim().len());
+            network_relay
+                .blocking_send(NetworkMsg::Broadcast(
+                    bincode::serialize(&Message {
+                        user,
+                        msg: input.as_bytes().to_vec().into_boxed_slice(),
+                    })
+                    .unwrap()
+                    .into_boxed_slice(),
+                ))
+                .unwrap();
+            tracing::debug!("[sending]: {}...", input);
+        });
+
+        // print received messages
+        while let Some(NetworkEvent::RawMessage(message)) = receiver.recv().await {
+            if let Ok(msg) = bincode::deserialize::<Message>(&message) {
+                if msg.user != user {
+                    println!(
```

Comment:
Maybe using [`tokio::io::stdout`](https://docs.rs/tokio/latest/tokio/io/fn.stdout.html) instead?

Reply:
Not sure the additional complexity is worth it since we can expect `println!` to be executed reasonably fast. 
`tokio::io:;stdout` may also result in interleaved output.

Blocking is also the default behavior for `tracing-subscriber` for the reasons outlined [here](https://github.com/tokio-rs/tracing/issues/668#issuecomment-609945964), so I think it makes sense to follow the same approach.
[This](https://www.fpcomplete.com/blog/different-levels-async-rust/) is also an interesting read.